server/src-lib/Hasura/GraphQL/Transport/HTTP.hs

```diff
@@ -42,24 +43,45 @@ import qualified Language.GraphQL.Draft.Syntax          as G
 import qualified Network.HTTP.Types                     as HTTP
 import qualified Network.Wai.Extended                   as Wai
 
+data QueryCacheKey = QueryCacheKey
+  { qckQueryString :: !GQLReqParsed
+  , qckUserRole    :: !RoleName
+  }
+
+instance J.ToJSON QueryCacheKey where
+  toJSON (QueryCacheKey qs ur ) =
+    J.object ["query_string" J..= qs, "user_role" J..= ur]
+
 
 class Monad m => MonadExecuteQuery m where
-  executeQuery
-    :: GQLReqParsed
-    -> [QueryRootField UnpreparedValue]
-    -> Maybe EQ.GeneratedSqlMap
-    -> PGExecCtx
-    -> TraceT (LazyTx QErr) EncJSON
-    -> TraceT (ExceptT QErr m) (HTTP.ResponseHeaders, EncJSON)
+  cacheLookup
+    :: [QueryRootField UnpreparedValue]
+    -- ^ Used to check that the query is cacheable
+    -> QueryCacheKey
+    -- ^ Key that uniquely identifies the result of a query execution
+    -> TraceT m (HTTP.ResponseHeaders, Maybe EncJSON)
```

Comment:
What's the case where response headers are returned from the cache but not the response?

Reply:
We've kept `cacheLookup` as a name, but it is a bit of a misnomer: it's more of a preprocessing steps; it *both* computes the response headers and does the cache lookup; we always use the response headers that it returns, regardless of whether there was a cache hit or not.